Vue-ify Upload Dialog w/small fixes

jmchilton opened the pull request:
See individual commits for details.

- Removed upload-button.js - mostly unused after #8368.
- Simplifications and ES6 stuff to existing upload Backbone components.
- Swap upload-view to UploadModal and use ``b-modal`` instead of Galaxy's modal stuff - use ``no-enforce-focus`` to ensure Select2 works within the modal.
- Migrate usage of upload Tabs to ``b-tabs`` - (initial shim for Backbone views in early commits replaced with Vue components for each tab type in subsequent commits).
- Replace each ``mvn/upload/<type>/<type>-view.js`` Backbone View with a Vue component.
   - Migrate composite-view.js to a new Composite.vue - eliminates Backbone view for everything but the row. Still relies of Backbone models.
   - Migrate collection-view.js to a new Collection.vue - eliminates Backbone view for everything but the row. Still relies of Backbone models.
   - Migrate rules-input-view.js to a new RulesInput.vue - eliminates Backbone view, never used row views or models.
   - Migrate default-view.js to a new Default.vue - eliminates Backbone view,  eliminates Backbone view for everything but the row rending. Still relies on Backbone models.
    - More reuse than Backbone code given mixins and components - this PR vastly reduces the duplication between the collection and default uploads.
- Parameterized lazy limited widget handling between Default.vue and Collection.vue - either can have a lazy limited set via a Vue property. Really reduces the divergence between default upload and collection upload nicely. I added some tests for this and fixed an off by one error inside of lazy-limited.js.
- Largely unrelated to Vue, but I took a first crack at eliminating a lot of the copy-paste duplication between the Backbone views in default-row.js and collection-row.js. Just cleaning up past messes I made ahead of potentially Vue-ifying these as well.
- Eliminates some qunit tests, adds more modern vuejs-utils tests. Not exactly one-to-one because some jquery plugins behave different in karma vs. qunit but definitely considerably more tests now.
- Generalize special case handling of FTP files that uploads them all in a single request for default uploads to work with collection uploads also - removing divergence between two upload modalities.

jmchilton committed:
Remove upload-button.js - absorbed by UploadButton.vue

UploadButton.vue added in https://github.com/galaxyproject/galaxy/pull/8368

jmchilton committed:
Simplify upload-view - no need to set unused properties.

jmchilton committed:
More cleanup of upload button plumbing.

jmchilton committed:
Migrate upload-view to new UploadModal.vue using b-modal

jmchilton committed:
Migrate upload tabs Backbone to Vue and b-tabs.

jmchilton committed:
Migrate upload composite-view to new Composite.vue

Fill out Select2.js as part of that.

jmchilton commented:
~~Continued the vue-ification throwing in the tabs and the actual composite uploader view... select2 still not working. I think the way the tabs work - this is all now creating the select2 components later... I had hope that would fix things.~~ (Fixed see comment below)

jmchilton commented:
It was the tabindex on b-modal... https://github.com/select2/select2-bootstrap-theme/issues/41

https://bootstrap-vue.js.org/docs/components/modal/ no-enforce-focus on the modal seems to fix it - but it is strongly discouraged. I guess it is no worse than our current modals?

jmchilton committed:
no-enforce-focus needed for select2 to work properly

jmchilton committed:
Migrate upload collection-view into new Collection.vue

jmchilton committed:
Progress on upload unit testing.

jmchilton committed:
Fix title...

jmchilton committed:
Fix anonymous users..

jmchilton committed:
Migrate upload rules-input-view to RulesInput.vue.

jmchilton committed:
Migrate upload default-view to new Default.vue.

Rework a lot of Collection.vue into shared mixin. There was a lot of duplication in there (all my fault).

jmchilton committed:
More unit tests for upload stuff.

jmchilton committed:
Add unit tests for UploadModal.

jmchilton committed:
Refactor upload collection/default row handling to share status handling.

Removes some divergence.

jmchilton committed:
Eliminate more duplication between default and collection rows.

jmchilton committed:
Eliminate lazy-limited differents between regular and collection upload.

- Let the components just think about this abstractly and move the logic into the common base class.
- Add tests!

jmchilton committed:
Fix off by one error in lazy-limited.js

jmchilton committed:
Merge _uploadFtp functionality between collection and default uploads.

jmchilton committed:
Rebase into Default.vue commit.

jmchilton commented:
@galaxybot test this

jmchilton committed:
Fix upload tab ids for Selenium tests.

guerler commented:
Noticed that the select2 dropdown width seems to be fixed to the actual parent dom width:

![widthselect2](https://user-images.githubusercontent.com/2105447/72915578-89dcb200-3d0e-11ea-9e00-70b49bd29d0b.png)

guerler commented on client/galaxy/scripts/components/Upload/Composite.vue:
```diff
@@ -0,0 +1,211 @@
+<template>
+    <upload-wrapper wrapperClass="upload-view-composite">
+        <div class="upload-helper" v-show="showHelper">Select extension to show files required to upload</div>
```
Maybe "Select Composite Type to show required files." instead since the select field is labeled: "Composite Type".

jmchilton replied:
Good catch - made the change.

jmchilton committed:
Allow Select2 dropdown width to exceed parent DOM width.

Seems to work for all components using it.

xref https://github.com/galaxyproject/galaxy/pull/9167#issuecomment-577283724

jmchilton committed:
Modify language for composite upload request.

xref https://github.com/galaxyproject/galaxy/pull/9167#pullrequestreview-346759554

jmchilton commented:
I've fixed the Select2 width issue with https://github.com/galaxyproject/galaxy/pull/9167/commits/224a92aaa290efb2ccf4061bc7db48ae5bd2e0d2 I believe. Thanks for the catch @guerler

guerler commented:
I think we need to vue-ify the backbone rows too, other than that this looks great.

jmchilton commented:
> I think we need to vue-ify the backbone rows too

Eventually that would make sense, but there is no reason that would have to happen as part of this PR I assume. I did this to enable #9110 and #9114 not to just to Vue for the sake of Vue-ing.

guerler committed:
Merge branch 'dev' into upload_vue_0

dannon commented:
This seems fine as an incremental step.  As we eliminate the rest of the backbone (especially the models..) it'll all be a lot cleaner.

Thanks!
